dlk/python/dlk/core/optimizer.py

```diff
@@ -14,632 +14,462 @@
-        """Prune the node and its inputs."""
-        # prune inputs
-        for i in node.input_ops.values():
-            if i not in self._nodes_removed:
-                self._prune(i)
-
-        # prune itself
-        self._nodes_removed.add(node)
-
-    def _precompute_or_prune_inputs(self, node: Operator) -> None:
-        """Precompute itself or prune the input nodes.
-
-        If all input has precompute value, then make the node precompute.
-        Otherwise, all prunable input nodes are pruned and substituted with
-        a new constant node.
-        """
-        ops: List[Operator] = [node.input_ops[i] for i in node.input_names if node.input_ops.get(i)]
-        ops_have_precomp_values = list(map(lambda x: self._has_precompute_value(x), ops))
-        ops_are_prunable = list(map(lambda x: self._is_prunable(x), ops))
-        ops_are_in_quantized = list(map(lambda x: x.name in self._quantizers.keys(), ops))
-
-        # check which input node can be pruned
-        if reduce(lambda x, y: x and y, ops_have_precomp_values):  # all input has concrete values
-            node.run_forward()
-            self._precomp_dic[node.name] = True  # this node can be pruned
-            if reduce(lambda x, y: x or y, ops_are_in_quantized):  # some input operator to be quantized exists
-                quantizers = {op.name: self._quantizers[op.name] for op in ops if self._quantizers.get(op.name)}
-                if len(quantizers) > 1:
-                    ValueError(f'{node.name}: multiple quantized inputs with {node.op_type} are not supported.')
-                self._quantizers[node.name] = list(quantizers.values())[0]
-
-        else:
-            self._precomp_dic[node.name] = False
+    done = False
+    processed_nodes = []
+    while not done:
+        p = Pattern('*')
+        matches = find_pattern(graph, p)
+        processed_before_precompute = len(processed_nodes)
+        to_be_removed = []
+
+        for m in matches:
+            if m.node in processed_nodes:
+                continue
+
+            # We want operators with inputs
+            if not m.node.input_nodes:
+                continue
+
+            precomputable = True
+            for input_node in m.node.input_nodes:
+                if input_node.op_type != 'Constant':
+                    precomputable = False
+
+            if not precomputable:
+                continue
+
+            processed_nodes += m.node.input_nodes
+            processed_nodes.append(m.node)
+
+            data = m.node.run_forward()
+
+            new_constant = Constant(
+                m.node.name + '_new',
+                m.node.dtype,
+                data,
+                dimension_format=m.node.dimension
+            )
+            graph.add_op(new_constant)
+
+            # get nodes to be removed after being disconnected
+            get_nodes_in_branch(m.node, None, to_be_removed)
+
+            new_constant.add_outputs({'output': m.node.output_ops.values()})
```

Comment:
Why these several lines are in here? looks general manupulation of graph and op.
How is to implement like `replace_op()` in `Graph` class and use `graph.replace_op(m.node, new_constant)` ?